src/server/coprocessor.rs

```diff
@@ -286,132 +277,195 @@ fn is_point(range: &KeyRange) -> bool {
-            return Ok(rows);
+struct SelectContext<'a> {
+    sel: SelectRequest,
+    snap: SnapshotStore<'a>,
+    eval: Evaluator,
+    cond_cols: HashMap<i64, ColumnInfo>,
+}
+
+fn collect_col_in_expr(cols: &mut HashMap<i64, ColumnInfo>,
+                       col_meta: &[ColumnInfo],
+                       expr: &Expr)
+                       -> Result<()> {
+    if expr.get_tp() == ExprType::ColumnRef {
+        let i = box_try!(number::decode_i64(expr.get_val()));
+        for c in col_meta {
+            if c.get_column_id() == i {
+                cols.insert(i, c.clone());
+                return Ok(());
+            }
         }
-        let h = box_try!(table::decode_handle(range.get_start()));
-        if let Some(row) = try!(get_row_by_handle(snap, sel, h, eval)) {
-            rows.push(row);
+        return Err(box_err!("column meta of {} is missing", i));
+    }
+    for c in expr.get_children() {
+        try!(collect_col_in_expr(cols, col_meta, c));
+    }
+    Ok(())
+}
+
+impl<'a> SelectContext<'a> {
+    fn new(sel: SelectRequest, snap: SnapshotStore<'a>) -> Result<SelectContext<'a>> {
+        let mut ctx = SelectContext {
+            sel: sel,
+            snap: snap,
+            eval: Default::default(),
+            cond_cols: Default::default(),
+        };
+        if !ctx.sel.has_field_where() {
+            return Ok(ctx);
         }
-    } else {
-        let mut seek_key = range.take_start();
-        loop {
-            trace!("seek {:?}", seek_key);
-            let mut res = try!(snap.scan(Key::from_raw(seek_key), 1));
-            if res.is_empty() {
-                debug!("no more data to scan.");
-                break;
+        {
+            let cols = if ctx.sel.has_table_info() {
+                ctx.sel.get_table_info().get_columns()
+            } else {
+                ctx.sel.get_index_info().get_columns()
+            };
+            try!(collect_col_in_expr(&mut ctx.cond_cols, cols, ctx.sel.get_field_where()));
+        }
+        Ok(ctx)
+    }
+
+    fn get_rows_from_sel(&mut self, ranges: Vec<KeyRange>) -> Result<Vec<Row>> {
+        let mut rows = vec![];
+        for ran in ranges {
+            let ran_rows = try!(self.get_rows_from_range(ran));
+            rows.extend(ran_rows);
+        }
+        Ok(rows)
+    }
+
+    fn get_rows_from_range(&mut self, mut range: KeyRange) -> Result<Vec<Row>> {
+        let mut rows = vec![];
+        if is_point(&range) {
+            if let None = try!(self.snap.get(&Key::from_raw(range.get_start().to_vec()))) {
+                return Ok(rows);
             }
-            let (key, _) = try!(res.pop().unwrap());
-            if range.get_end() <= &key {
-                debug!("reach end key: {:?} >= {:?}", key, range.get_end());
-                break;
+            let h = box_try!(table::decode_handle(range.get_start()));
+            if try!(self.should_skip(h)) {
```

Comment:
`should_skip` seems a little confused,  it looks like we skip this handle and jump to a new handle in this range.